### PR TITLE
Add EventBridge Pipes blog post

### PR DESCRIPTION
**Pull request by rrhodes**
# What?

Blog post about use of EventBridge Pipes to share power updates on the KF platform.

### CODE REVIEW

File: _posts/2024-02-14-using-eventbridge-pipes.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+author: Ross Rhodes
+banner: /assets/img/posts/2015-11-23-tech-jobs.jpg
+hex: 0e1720
+---
+
+Here at Kraken, we host a Dispatch API to read and write power dispatches. This API forms a critical part of the wider KrakenFlex platform, interacting with other internal APIs to control power assets. In recent months, the number of requests processed by the API has significantly increased. We expect this trend to persist as we integrate more batteries, electric vehicles, and other assets onto the platform. If left unaddressed, this could lead to performance degradation under the API’s current architecture.
+
+With that in mind, the Control & Dispatch team developed a new API - the Power API - which is designed to read and write dispatches for more assets whilst delivering performance improvements when processing requests. The team's worked hard on the development of Power API, which will soon replace the Dispatch API.
+
+<img src="/assets/img/posts/2024-02-14-using-eventbridge-pipes/system-change-notification-banner.png" width="725" />
```
Comment (rrhodes) [inline _posts/2024-02-14-using-eventbridge-pipes.md:14]: I don't think previous blog posts have shared images throughout their posts, but correct me if I'm mistaken. Happy to take suggestions if there's another preferred way to share pictures.

File: _posts/2024-02-14-using-eventbridge-pipes.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
+---
+title: Sharing Power Updates using Amazon EventBridge Pipes
+layout: post
+category: news
+author: Ross Rhodes
+banner: /assets/img/posts/2015-11-23-tech-jobs.jpg
+hex: 0e1720
+---
+
+Here at Kraken, we host a Dispatch API to read and write power dispatches. This API forms a critical part of the wider KrakenFlex platform, interacting with other internal APIs to control power assets. In recent months, the number of requests processed by the API has significantly increased. We expect this trend to persist as we integrate more batteries, electric vehicles, and other assets onto the platform. Over time, if no action is taken, continued increases will lead to performance degradation under the API’s current architecture.
```
[the remark this answers is not in the document]
Comment (frederizia) [reply _posts/2024-02-14-using-eventbridge-pipes.md:10]: And generally what is meant by "powers" which is referred to a lot (e.g. "new and updated powers" further below - as someone not familiar with KF I don't really understand what is meant by that).

File: _posts/2024-02-14-using-eventbridge-pipes.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+After exploring a few options, we settled on a relatively new service from AWS called [EventBridge Pipes](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes.html). Pipes ease the integration process between AWS resources — reducing the need for Lambda functions hosting custom application code. For our use case, they support a direct integration between DynamoDB Streams and EventBridge Event Buses.
+
+<img src="/assets/img/posts/2024-02-14-using-eventbridge-pipes/eventbridge-pipe-architecture.png" width="725" />
+
+At a high level, Pipes consist of four stages:
+
+1. Source AWS service
+1. Filtering of data (optional)
+1. Data enhancement (optional)
+1. Target AWS service
```
Comment (frederizia) [inline _posts/2024-02-14-using-eventbridge-pipes.md:38]: 1.1.1.1.?

File: _posts/2024-02-14-using-eventbridge-pipes.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+          Filters:
+            - Pattern: '{ "eventName": ["INSERT", "MODIFY"] }'
+      Target: !GetAtt EventPublisherEventBus.Arn
+      TargetParameters:
+        InputTemplate: '{ "subjectId": <$.dynamodb.NewImage.subjectId.S> }'
+```
+
+Note that other resources are omitted from the snippet. For example, the DynamoDB table `PowersTable` and the Event Bus `EventPublisherEventBus`. Furthermore, it is possible to transform input data for the target and enhancement steps. In the case of Power API, Stream record data nested within `dynamodb` → `NewImage` is extracted to simplify what the Event Bus receives.
+
+With a Pipe provisioned and successfully forwarding Stream records to the Event Bus, further consideration was given regarding what happens in error scenarios. A few questions emerged:
```
Comment (frederizia) [inline _posts/2024-02-14-using-eventbridge-pipes.md:69]: Just a small note that the tense that you write in is a little inconsistent. I get that above you are describing the current state ("we are doing this") but then here you're talking about it more along the lines of "this was the decision making and set up process we followed" (in the past tense) which is a bit confusing/the mix doesn't read as well.

File: _posts/2024-02-14-using-eventbridge-pipes.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+* What happens if the Pipe fails to put records onto the Event Bus?
+* How long do we have to remediate problems when they arise?
+* How are we made aware of Pipe errors when they occur?
+
+## Error Handling With Pipes
```
Comment (frederizia) [inline _posts/2024-02-14-using-eventbridge-pipes.md:75]: I would make sub sections sentence case not title case. Might be a personal preference but I think this is what we've done in previous blogs as well (I just spot checked).

File: _posts/2024-02-14-using-eventbridge-pipes.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+We observe application behavior through [CloudWatch Metrics](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/working_with_metrics.html) and we use [Grafana](https://grafana.com/) dashboards to ease observability. Pipes offer a [range of metrics](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-monitoring.html) to monitor invocations, processing durations, and execution errors. We also [contributed to Grafana](https://github.com/grafana/grafana/pull/69994) to make it easier to add Pipe metrics to dashboards. Below is a screenshot of our Pipe dashboard panels.
```
Comment (frederizia) [inline _posts/2024-02-14-using-eventbridge-pipes.md:77]: Love the shout out to the Grafana contribution 💪

File: _posts/2024-02-14-using-eventbridge-pipes.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+Even if we were to address that particular issue, there is no AWS managed re-drive mechanism at the time of writing to move Stream records from the DLQ back to the Pipe. There’s also no API or SDK available to trigger the Pipe with a record. There's a couple approaches we considered to work around these limitations. One option requires custom logic within a Lambda function to read from the DLQ, replicate the Pipe’s filtering and transformation logic, then send records directly to the Event Bus. This is considerable overhead for us to manage, which itself could be error-prone.
```
Comment (frederizia) [inline _posts/2024-02-14-using-eventbridge-pipes.md:87]: 🐼 "There are a couple of approaches"

File: _posts/2024-02-14-using-eventbridge-pipes.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+      Namespace: AWS/EventBridge/Pipes
+      Period: 60 # seconds
+      Statistic: Sum
+      Threshold: 0
+      TreatMissingData: notBreaching
+```
+
+## Conclusion
+
+We’re satisfied Pipes are a great resource to forward DynamoDB Stream records to an EventBridge Event Bus. After we originally drafted this blog post, the Control & Dispatch  team migrated from a DynamoDB Stream to a [Kinesis Data Stream](https://aws.amazon.com/kinesis/data-streams/) to capture power changes, which allows us to set up more than two stream consumers without observing an impact on stream read performance.
```
Comment (frederizia) [inline _posts/2024-02-14-using-eventbridge-pipes.md:129]: There's a roque space after "Control& Dispatch".

I'm not sure the reader needs to know what happened between drafting this and actually publishing - can this information live elsewhere? Or does this mean the blog is a bit out of date already?

**Review by frederizia — COMMENTED** [review]
Nice post! I like that you included configuration examples which might help someone else get started.

I think setting the scene at the start could be improved a little. As someone who has very little background on what Kraken Flex does, some of the terminology could be explained a bit first. I know this is more about the technical implementation but it would be nice to explain the context a bit more.

File: _posts/2024-02-14-using-eventbridge-pipes.md (hunk 3 above)
Comment (rrhodes) [reply _posts/2024-02-14-using-eventbridge-pipes.md:38]: This is Markdown which generates a numbered list starting from 1. We don't need to manually increment numbers ourselves.

**Comment by rrhodes** [comment]
Hey @frederizia, thanks for the thorough feedback! Let me review this later today / tomorrow and give you a shout when a second draft is ready.

File: _posts/2024-02-14-using-eventbridge-pipes.md (hunk 8 above)
Comment (rrhodes) [reply _posts/2024-02-14-using-eventbridge-pipes.md:129]: No problem omitting this part of the blog post - there's plenty to discuss there that could form its own post another day. 👍🏻

**Comment by rrhodes** [comment]
Hi @frederizia, that should be everything addressed now. Let me know your thoughts. 🙂

**Comment by rrhodes** [comment]
Example screenshot of how the images render.

**Comment by rrhodes** [comment]
Example screenshot of how the images render.

Screenshot 2024-02-15 at 09 17 27

File: Gemfile [hunk 9]
```diff
@@ -2,3 +2,5 @@ source 'https://rubygems.org'
 
 # See https://pages.github.com/versions/ for github-pages version that GH uses.
 gem 'github-pages', "~> 228", group: :jekyll_plugins
+
+gem "webrick", "~> 1.8"
```
[the remark this answers is not in the document]
Comment (rrhodes) [reply Gemfile:6]: @codeinthehole what're your thoughts on this?

File: _posts/2024-02-14-using-eventbridge-pipes.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (rrhodes) [reply _posts/2024-02-14-using-eventbridge-pipes.md]: Good shout. I've elaborated on the meaning of "powers" and "assets" at the start of the blog post.

File: _posts/2024-02-14-using-eventbridge-pipes.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+layout: post
+category: news
+author: Ross Rhodes
+banner: /assets/img/posts/2015-11-23-tech-jobs.jpg
+hex: 0e1720
+---
+
+Here at Kraken, we host a Dispatch API to read and write power dispatches. This API forms a critical part of the wider KrakenFlex platform, interacting with other internal APIs to control power assets. In recent months, the number of requests processed by the API has significantly increased. We expect this trend to persist as we integrate more batteries, electric vehicles, and other assets onto the platform. Over time, if no action is taken, continued increases will lead to performance degradation under the API’s current architecture.
+
+With that in mind, the Control & Dispatch team developed a new API - the Power API - which is designed to read and write powers (including dispatches) for more assets whilst delivering performance improvements. The team's worked hard on the development of Power API, which will soon replace the Dispatch API.
```
[the remark this answers is not in the document]
Comment (rrhodes) [reply _posts/2024-02-14-using-eventbridge-pipes.md]: Noted, I've clarified this point.

File: _posts/2024-02-14-using-eventbridge-pipes.md (hunk 4 above)
Comment (rrhodes) [reply _posts/2024-02-14-using-eventbridge-pipes.md]: Good call, I think it's important to get this right. I've amended this blog post so it should use present tense throughout.

File: _posts/2024-02-14-using-eventbridge-pipes.md (hunk 5 above)
Comment (rrhodes) [reply _posts/2024-02-14-using-eventbridge-pipes.md]: Noted, happy to keep this consistent with other posts.

File: _posts/2024-02-14-using-eventbridge-pipes.md (hunk 7 above)
Comment (rrhodes) [reply _posts/2024-02-14-using-eventbridge-pipes.md]: Good spot, resolved.

File: _posts/2024-02-16-using-eventbridge-pipes.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+<img src="/assets/img/posts/2024-02-16-using-eventbridge-pipes/serverless-architecture.png" width="725" />
+
+Several internal services were already subscribed to dispatch update events broadcast by the Dispatch API. Therefore we required Power API to emit equivalent events before we could migrate these services away from Dispatch API. We adopted [Amazon EventBridge](https://aws.amazon.com/eventbridge/) to fulfil this requirement using an [Event Bus](https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-event-bus.html), which acts as a broker between
```
Comment (frederizia) [inline _posts/2024-02-16-using-eventbridge-pipes.md:20]: 🐼 🐼 I think there should be a comma after therefore (at least pycharm always tells me to add one).

Comment (frederizia) [reply _posts/2024-02-16-using-eventbridge-pipes.md:20]: There's another "therefore" further down.

**Review by frederizia — APPROVED** [review]
I've got some more minor comments but I think this is looking good! Would still wait for Wints to comment on the requirements addition as I don't really know much about ruby.

File: _posts/2024-02-16-using-eventbridge-pipes.md [hunk 12]
```diff
@@ -0,0 +1,131 @@
+---
+title: Sharing Power Updates using Amazon EventBridge Pipes
+layout: post
+category: news
+author: Ross Rhodes
+banner: /assets/img/posts/2015-11-23-tech-jobs.jpg
+hex: 0e1720
+---
+
+Within the Kraken platform, we host a Dispatch API responsible for reading and writing dispatches on assets. A dispatch is an import or export of power, whilst an asset can take many different forms: ranging from commercial-scale batteries to electric vehicles. In recent months, the number of dispatch requests processed by the API has significantly increased. We expect this trend to persist as more assets integrate onto the platform. Over time, without taking action, continued increases would lead to performance degradation under the API’s current architecture.
```
[the remark this answers is not in the document]
Comment (rrhodes) [reply _posts/2024-02-16-using-eventbridge-pipes.md:10]: This was raised with me in my initial feedback last year. I was told to use Kraken Tech, where previously I was using KrakenFlex.

File: _posts/2024-02-16-using-eventbridge-pipes.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+<img src="/assets/img/posts/2024-02-16-using-eventbridge-pipes/eventbridge-pipe-architecture.png" width="725" />
+
+At a high level, Pipes consist of four stages:
+
+1. Source AWS service
+1. Filtering of data (optional)
+1. Data enhancement (optional)
+1. Target AWS service
+
+For Power API the source integration is the DynamoDB Stream, and the target is the Event Bus. Filtering is configurable to specify which Stream records to put onto the Event Bus. While the Stream captures DynamoDB record inserts, updates, and deletions, we filter out deletions at this stage. We also skip the enhancement stage, which would require us to add additional infrastructure and / or custom logic within Lambda functions. The raw data from the Stream suffices for our needs.
```
[the remark this answers is not in the document]
Comment (rrhodes) [reply _posts/2024-02-16-using-eventbridge-pipes.md]: Good spot, resolved!

File: _posts/2024-02-16-using-eventbridge-pipes.md (hunk 11 above)
Comment (rrhodes) [reply _posts/2024-02-16-using-eventbridge-pipes.md]: Added commas after each "therefore".

File: Gemfile (hunk 9 above)
[the remark this answers is not in the document]
Comment (codeinthehole) [reply Gemfile:6]: Am no Ruby expert. Happy for this to go in.

**Review by codeinthehole — APPROVED** [review]

**Pull request merged**